lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,653 @@
+//===- GrandCentralTaps.cpp -------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the GrandCentralTaps pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "./PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/Support/Allocator.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/FormatVariadic.h"
+
+#define DEBUG_TYPE "gct"
+
+using namespace circt;
+using namespace firrtl;
+using mlir::function_like_impl::getArgAttrDict;
+using mlir::function_like_impl::setAllArgAttrDicts;
+
+//===----------------------------------------------------------------------===//
+// Utilities
+//===----------------------------------------------------------------------===//
+
+/// An absolute instance path.
+using InstancePath = ArrayRef<InstanceOp>;
+
+template <typename T>
+static T &operator<<(T &os, const InstancePath &path) {
+  os << "$root";
+  for (auto inst : path)
+    os << "." << inst.name();
+  return os;
+}
+
+namespace {
+
+/// A port annotated with a data tap key or mem tap.
+struct AnnotatedPort {
+  unsigned portNum;
+  Annotation anno;
+};
+
+/// An extmodule that has annotated ports.
+struct AnnotatedExtModule {
+  FExtModuleOp extModule;
+  SmallVector<AnnotatedPort, 4> portAnnos;
+  // Module annotations without data tap stuff.
+  ArrayAttr filteredModuleAnnos;
+  /// Port annotations without data tap stuff.
+  SmallVector<AnnotationSet, 8> filteredPortAnnos;
+};
+
+/// A value annotated to be tapped.
+struct TappedValue {
+  Value value;
+  Annotation anno;
+};
+
+/// A data structure that tracks the instances of modules and can provide
+/// absolute paths to these instances.
+struct InstanceGraph {
+  /// The root circuit.
+  CircuitOp circuitOp;
+  /// The main module in the circuit.
+  Operation *mainModule;
+  /// A mapping from a module to all its instances in the design.
+  DenseMap<Operation *, SmallVector<InstanceOp, 1>> moduleInstances;
```

Comment:
instead of `SmallVector<InstanceOp, 1>`, you might consider llvm/ADT/TinyPtrVector.h.  It is sizeof(one word) instead of sizeof(4 words)

Reply:
Nice, didn't know about that container! Changed.